### PR TITLE
Added support for native fence android extension

### PR DESCRIPTION
**Pull request by emilwestergren**
@erikhagglund peer review

### CODE REVIEW

**Comment by marcusnaslund** [comment]
PRInt fails everything.

File: source/draw/gpu/opengl/backend/EGLNativeFence.ooc [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+import egl/egl
+import GLExtensions, GLFence
+version(!gpuOff) {
+EGLNativeFence: class extends GLFence {
+	_display: Pointer
+	init: func (=_display) { super() }
+	free: override func {
+		GLExtensions eglDestroySyncKHR(this _display, this _backend)
```
Comment (erikhagglund) [inline source/draw/gpu/opengl/backend/EGLNativeFence.ooc]: Check return value and add "validateStart/validateEnd"

File: source/draw/gpu/opengl/backend/EGLNativeFence.ooc [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+import egl/egl
+import GLExtensions, GLFence
+version(!gpuOff) {
+EGLNativeFence: class extends GLFence {
+	_display: Pointer
```
Comment (erikhagglund) [inline source/draw/gpu/opengl/backend/EGLNativeFence.ooc]: This should be of type EGLDisplay, but that should be all over the code i guess so does not belong to this PR.

File: source/draw/gpu/opengl/backend/EGLNativeFence.ooc [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+		GLExtensions eglDestroySyncKHR(this _display, this _backend)
+		super()
+	}
+	sync: override func {
+		this _backend = GLExtensions eglCreateSyncKHR(this _display, EGL_SYNC_NATIVE_FENCE_ANDROID, null as Int*)
+	}
+	clientWait: override func (timeout: ULong = ULong maximumValue) -> Bool {
+		GLExtensions eglClientWaitSyncKHR(this _display, this _backend, EGL_SYNC_FLUSH_COMMANDS_BIT_KHR, timeout)
+	}
+	wait: override func { raise("Wait unimplemented for EGLNativeFence") }
```
Comment (erikhagglund) [inline source/draw/gpu/opengl/backend/EGLNativeFence.ooc]: Was not Debug error preferred?

File: source/draw/gpu/opengl/backend/EGLNativeFence.ooc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	init: func (=_display) { super() }
+	free: override func {
+		GLExtensions eglDestroySyncKHR(this _display, this _backend)
+		super()
+	}
+	sync: override func {
+		this _backend = GLExtensions eglCreateSyncKHR(this _display, EGL_SYNC_NATIVE_FENCE_ANDROID, null as Int*)
+	}
+	clientWait: override func (timeout: ULong = ULong maximumValue) -> Bool {
+		GLExtensions eglClientWaitSyncKHR(this _display, this _backend, EGL_SYNC_FLUSH_COMMANDS_BIT_KHR, timeout)
```
Comment (erikhagglund) [inline source/draw/gpu/opengl/backend/EGLNativeFence.ooc]: no need for glFlush as in Gles3Fence wait()?

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/EGLNativeFence.ooc]: This is `clientWait`. The flag `EGL_SYNC_FLUSH_COMMANDS_BIT_KHR` will automatically flush.

File: source/draw/gpu/opengl/backend/EGLNativeFence.ooc (hunk 2 above)
Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/EGLNativeFence.ooc]: Yeah, we didn't care about creating ooc types for the EGL types since they are typedefs.

**Comment by emilwestergren** [comment]
This pull request has now been updated.

File: source/draw/gpu/opengl/backend/GLExtensions.ooc [hunk 5]
```diff
@@ -14,12 +14,24 @@ GLExtensions: class {
...
 		if (!This _initialized) {
 			This eglCreateImageKHR = This _load("eglCreateImageKHR") as Func(Pointer, Pointer, UInt, Pointer, Int*) -> Pointer
 			This eglDestroyImageKHR = This _load("eglDestroyImageKHR") as Func(Pointer, Pointer)
 			This glEGLImageTargetTexture2DOES = This _load("glEGLImageTargetTexture2DOES") as Func(UInt, Pointer)
+			This eglCreateSyncKHR = This _load("eglCreateSyncKHR") as Func(Pointer, UInt, Int*) -> Pointer
+			This eglDestroySyncKHR = This _load("eglDestroySyncKHR") as Func (Pointer, Pointer) -> Bool
+			This eglClientWaitSyncKHR = This _load("eglClientWaitSyncKHR") as Func (Pointer, Pointer, Int, ULong) -> Bool
+			version(!windows) {
+				//For some reason this function can't be loaded with eglGetProcAddress so we load it with dlsym instead
+				RTLD_DEFAULT := null //This should be defined in dlfcn.h
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: Should be, but isn't? Or what?

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: It's only defined when the flag GNU_SOURCE is defined for the source code. It works when I build vidhance, but not when only running the tests in kean.

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: Sorry, was just about to edit... :)

Comment (marcusnaslund) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: That's weird.

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]:
```c
#ifdef __USE_GNU
/* If the first argument of `dlsym' or `dlvsym' is set to RTLD_NEXT
   the run-time address of the symbol called NAME in the next shared
   object is returned.  The "next" relation is defined by the order
   the shared objects were loaded.  */
# define RTLD_NEXT	((void *) -1l)

/* If the first argument to `dlsym' or `dlvsym' is set to RTLD_DEFAULT
   the run-time address of the symbol called NAME in the global scope
   is returned.  */
# define RTLD_DEFAULT	((void *) 0)
```

Comment (marcusnaslund) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: No proble, thanks. :)

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: I also read `__USE_GNU` shouldn't be defined explicitly, but you should `#define _GNU_SOURCE`.

Comment (marcusnaslund) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: Yes, I remember something similar from when I sorted out the mess of external C includes in this repo.

If this works satisfactory then let's be happy with this. I've made an issue #1843 in case we have time to think about it further later.

Comment (emilwestergren) [inline source/draw/gpu/opengl/backend/GLExtensions.ooc]: It works because we know the definition of `RTLD_DEFAULT` is just `null`, but it's ugly. I think it's fine for now but it should be fixed.

**Comment by marcusnaslund** [comment]
Doesn't build for android.

**Comment by emilwestergren** [comment]
Yes, I think this needs the latest NDK version

**Comment by marcusnaslund** [comment]
Let's have @fredrikbryntesson install the latest NDK on PRInt when he gets back.

**Comment by marcusnaslund** [comment]
`EGLNativeFence.ooc:25:218: error: 'EGL_SYNC_NATIVE_FENCE_ANDROID' undeclared` even though we are using the updated NDK now.

**Comment by emilwestergren** [comment]
Are you sure? When I try to build locally, it fails with that exact error when using old NDK, but as soon as I change to the latest version it builds.

**Comment by marcusnaslund** [comment]
> Are you sure?

Nothing is ever for sure with PRInt. I will double-check tomorrow.

**Comment by marcusnaslund** [comment]
Ohh, I see the problem now. The make-script for android in `develop` still points to `android-ndk-r11c`. It's only the script in `<secret feature>1.5` which points to `r12b`. How to do this?

**Comment by marcusnaslund** [comment]
Frozen awaiting release.

**Comment by emilwestergren** [comment]
I can create this against `primaryFeature1.5` instead I guess?

**Comment by marcusnaslund** [comment]
Sure.

**Comment by marcusnaslund** [comment]
Frozen awaiting release.

**Comment by marcusnaslund** [comment]
Please rebase this, and we should be fine now. @emilwestergren

**Comment by emilwestergren** [comment]
Updated

**Comment by marcusnaslund** [comment]
That's a lot (17) of commits. I won't hold this PR back for it, but just make sure there's nothing "extra" accidentally in there.

**Comment by marcusnaslund** [comment]
Yep, now it works.
Please review again @erikhagglund

**Comment by erikhagglund** [comment]
Looks good to me, does not seem like any other code is trying to piggyback on this PR

**Comment by marcusnaslund** [comment]
Well then, after almost two months of waiting, this PR is getting merged.
